src/artrace.py

```diff
@@ -1350,159 +1054,144 @@ def model_tilt(slf, det, msarc, guesstilts=None, censpec=None, plotQA=False, ref
+        ofit = slf._argflag['trace']['orders']['pcatilt']
+        lnpc = len(ofit)-1
+        # Only do a PCA if there are enough good orders
+        if np.sum(1.0-extrap_row) > ofit[0]+1:
+            # Perform a PCA on the tilts
+            msgs.info("Performing a PCA on the tilts")
+            ordsnd = np.linspace(0.0, 1.0, msarc.shape[0])
+            xcen = xv[:, np.newaxis].repeat(msarc.shape[0], axis=1)
+            fitted, outpar = arpca.basis(xcen, tiltval, tcoeff, lnpc, ofit, weights=None,
+                                         x0in=ordsnd, mask=maskrw, skipx0=False,
+                                         function=slf._argflag['trace']['orders']['function'])
+            arpca.pc_plot(slf, outpar, ofit, pcadesc="Spectral Tilts PCA", addOne=False)
+            # Extrapolate the remaining orders requested
+            orders = np.linspace(0.0, 1.0, msarc.shape[0])
+            extrap_tilt, outpar = arpca.extrapolate(outpar, orders, function=slf._argflag['trace']['orders']['function'])
+            polytilts = extrap_tilt.T
+        else:
+            # Fit the model with a 2D polynomial
+            msgs.warn("Could not perform a PCA when tracing the spectral tilt" + msgs.newline() +
+                      "Not enough well-traced arc lines")
+            msgs.info("Fitting tilts with a low order, 2D polynomial")
+            wgd = np.where(xtilt != maskval)
+            coeff = arutils.polyfit2d_general(xtilt[wgd], ytilt[wgd], mtilt[wgd], fitxy)
+            polytilts = arutils.polyval2d_general(coeff, np.linspace(0.0, 1.0, msarc.shape[1]),
+                                                  np.linspace(0.0, 1.0, msarc.shape[0]))
+
+        # Now apply the starting guess for the tilts
         if guesstilts is not None:
-#            wfit = np.where(wtilt!=maskval)
-#            coeff = arutils.polyfit2d(xtilt[wfit], ytilt[wfit], ztilt[wfit], fitxy)#, w=wtilt[wfit]/np.max(wtilt[wfit]))
-#            tilts = arutils.polyval2d(coeff, np.arange(msarc.shape[1])/(msarc.shape[1]-1.0), np.arange(msarc.shape[0])/(msarc.shape[0]-1.0))
-            tilts += guesstilts
-    else:
-        msgs.error("Please use trace+tilts+fit1d")
-
-    # Perform a high order 2D polynomial fit to the best PCA tilts, then interpolate
-#    if refine_tilts:
-#        debugger.set_trace()
-
-    #xx, yy = np.meshgrid(np.arange(msarc.shape[1])/(msarc.shape[1]-1.0), np.arange(msarc.shape[0])/(msarc.shape[0]-1.0))
-    #coeff = arutils.polyfit2d(xx, yy, tilts, [8,8])
-    #tilts = arutils.polyval2d(coeff, np.arange(msarc.shape[1])/(msarc.shape[1]-1.0), np.arange(msarc.shape[0])/(msarc.shape[0]-1.0))
-    # Plot the model tilts
-
-#guesstilts[arcdet[j],ordcen[arcdet[j],0]-k]*float(msarc.shape[0]-1.0)
-
-
-    tiltsplot = tilts[arcdet,:].T
+            polytilts += guesstilts
+
+        if slf._argflag['trace']['orders']['tilts'].lower() in ["spline", "interp"]:
+            xspl = np.linspace(0.0, 1.0, msarc.shape[1])
```

Comment:
Agreed, but perhaps we should just fit the original xtilt, ytilt, and ztilt
values. At the moment, the spline is fit to the model of the tilts, but we
should be fitting the spline to the measured tilts (i.e. the original
result of the cross correlation). I'll look into this shortly...

On 29 April 2016 at 04:54, J. Xavier Prochaska <notifications@github.com>
wrote:

> In src/artrace.py
> <https://github.com/PYPIT/PYPIT/pull/49#discussion_r61565690>:
>
> > -#    if refine_tilts:
> > -#        debugger.set_trace()
> > -
> > -    #xx, yy = np.meshgrid(np.arange(msarc.shape[1])/(msarc.shape[1]-1.0), np.arange(msarc.shape[0])/(msarc.shape[0]-1.0))
> > -    #coeff = arutils.polyfit2d(xx, yy, tilts, [8,8])
> > -    #tilts = arutils.polyval2d(coeff, np.arange(msarc.shape[1])/(msarc.shape[1]-1.0), np.arange(msarc.shape[0])/(msarc.shape[0]-1.0))
> > -    # Plot the model tilts
> > -
> > -#guesstilts[arcdet[j],ordcen[arcdet[j],0]-k]*float(msarc.shape[0]-1.0)
> > -
> > -
> > -    tiltsplot = tilts[arcdet,:].T
> > +            polytilts += guesstilts
> > +
> > +        if slf._argflag['trace']['orders']['tilts'].lower() in ["spline", "interp"]:
> > +            xspl = np.linspace(0.0, 1.0, msarc.shape[1])
>
> I think this should be over a limited x range, not the full detector,
> e.g.
> [xint:xint+2*sz+1]
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/PYPIT/PYPIT/pull/49/files/7f6e721a24457cd7164550e1e1677a27d8702468#r61565690>
>